demos/slack-alert-app/go/slackbot.go

```diff
@@ -0,0 +1,171 @@
+	// HTTP error (>4xxx) count for each service in the `px-sock-shop` namespace.
+	// To deploy the px-sock-shop demo, see:
+	// https://docs.pixielabs.ai/tutorials/slackbot-alert for how to
+	pxlScript = `
+import px
+
+df = px.DataFrame(table='http_events', start_time='-5m')
+
+# Add column for HTTP response status errors.
+df.error = df.http_resp_status >= 400
+
+# Add columns for service, namespace info
+df.namespace = df.ctx['namespace']
+df.service = df.ctx['service']
+
+# Filter for px-sock-shop namespace only.
+df = df[df.namespace == 'px-sock-shop']
+
+# Group HTTP events by service, counting errors and total HTTP events.
+df = df.groupby(['service']).agg(
+    error_count=('error', px.sum),
+    total_requests=('http_resp_status', px.count)
+)
+
+px.display(df, "http_table")
+`
+)
+
+func main() {
+
+	// The slackbot requires the following configs, which are specified
+	// using environment variables. For directions on how to find these
+	// config values, see: https://docs.pixielabs.ai/tutorials/slackbot-alert
+	pixieAPIKey, ok := os.LookupEnv("PIXIE_API_KEY")
+	if !ok {
+		panic("Please set PIXIE_API_KEY environment variable.")
+	}
+
+	pixieClusterID, ok := os.LookupEnv("PIXIE_CLUSTER_ID")
+	if !ok {
+		panic("Please set PIXIE_CLUSTER_ID environment variable.")
+	}
+
+	slackToken, ok := os.LookupEnv("SLACK_BOT_TOKEN")
+	if !ok {
+		panic("Please set SLACK_BOT_TOKEN environment variable.")
+	}
+
+	ctx := context.Background()
+	pixieClient, err := pxapi.NewClient(ctx, pxapi.WithAPIKey(pixieAPIKey))
+	if err != nil {
+		panic(err)
+	}
+	vz, err := pixieClient.NewVizierClient(ctx, pixieClusterID)
+	if err != nil {
+		panic(err)
+	}
+
+	slackClient := slack.New(slackToken)
+
+	// Using a ticker (rather than a for loop + sleep call) means that the slackbot
+	// will message every 5 minutes, regardless of how long it takes to call the 
+	// Pixie API (as long as it takes less than 5 minutes).
+	ticker := time.NewTicker(5 * time.Minute)
+	for {
+		tm := &tableMux{tables: make(map[string]*tableCollector)}
+		log.Println("Executing PxL script.")
+		resultSet, err := vz.ExecuteScript(ctx, pxlScript, tm)
+		if err != nil && err != io.EOF {
+			panic(err)
+		}
+
+		log.Println("Stream PxL script results.")
+		if err := resultSet.Stream(); err != nil {
+			log.Printf("Got error: %+v, while streaming.\n", err)
+		}
+
+		// Get slack message constructed from table data.
+		table := tm.GetTable("http_table").GetTableDataSync()
+		log.Println("Sending slack message.")
+		_, _, err = slackClient.PostMessage(slackChannel, slack.MsgOptionText(table, false), slack.MsgOptionAsUser(true))
+		if err != nil {
+			log.Println("Error sending to slack: " + err.Error())
+		}
+
+		// wait for next tick
+		<-ticker.C
+	}
+}
+
+// Implement the TableRecordHandler interface to processes the PxL script table record-wise.
+type tableCollector struct {
+	msg strings.Builder
+	// Channel to communicate to tableMux that tableCollector is done collecting table rows.
```

Comment:
We aren't really communicating to tableMux. We are just holding that info and using it to block on a call to GetTableDataSync